// react-create-context/dist/react-create-context.esm.js
import React, { useEffect, createContext, useReducer, useContext, useState } from 'react';

var isServer = typeof window === 'undefined';

var useLocalStorageSync = function useLocalStorageSync(state, key) {
  useEffect(function () {
    if (key && !isServer) {
      window.localStorage.setItem(key, JSON.stringify(state));
    }
  }, [state, key]);
};

var getFromLocalStorage = function getFromLocalStorage(key) {
  if (!key || isServer) return null;
  var value = window.localStorage.getItem(key);
  if (value) return JSON.parse(value);
  return null;
};

var createContextStore = function createContextStore(storeReducer, initialState, key) {
  var localStorageState = getFromLocalStorage(key);
  var initialOrSavedState = localStorageState != null ? localStorageState : initialState;
  var DispatchContext = createContext({});
  var StoreContext = createContext(initialOrSavedState);

  var StoreProvider = function StoreProvider(_ref) {
    var children = _ref.children;

    var _useReducer = useReducer(storeReducer, initialOrSavedState),
        store = _useReducer[0],
        dispatch = _useReducer[1];

    useLocalStorageSync(store, key);
    return React.createElement(DispatchContext.Provider, {
      value: dispatch
    }, React.createElement(StoreContext.Provider, {
      value: store
    }, children));
  };

  var useDispatch = function useDispatch() {
    return useContext(DispatchContext);
  };

  var useStore = function useStore() {
    return useContext(StoreContext);
  };

  return [StoreProvider, useStore, useDispatch];
};

var createContextState = function createContextState(initialState, key) {
  var localStorageState = getFromLocalStorage(key);
  var initialOrSavedState = localStorageState != null ? localStorageState : initialState;
  var SetStateContext = createContext({});
  var StateContext = createContext(initialOrSavedState);

  var StateProvider = function StateProvider(_ref) {
    var children = _ref.children;

    var _useState = useState(initialOrSavedState),
        state = _useState[0],
        setState = _useState[1];

    useLocalStorageSync(state, key);
    return React.createElement(SetStateContext.Provider, {
      value: setState
    }, React.createElement(StateContext.Provider, {
      value: state
    }, children));
  };

  var useContextSetState = function useContextSetState() {
    return useContext(SetStateContext);
  };

  var useContextState = function useContextState() {
    return useContext(StateContext);
  };

  return [StateProvider, useContextState, useContextSetState];
};

export { createContextState, createContextStore };
